### ISSUE TITLE
state example insufficiently explained

### ISSUE THREAD

**Issue by gitgithan** [marimo-team/marimo#2738]
### Documentation is

- [ ] Missing
- [ ] Outdated
- [x] Confusing
- [ ] Not sure?

### Explain in Detail

I can't understand the dataflow of below example from https://docs.marimo.io/guides/state.html and also `marimo tutorial ui` 

I have many questions in bold below.

```
def add_task():
    if task_entry_box.value:
        set_tasks(lambda v: v + [Task(task_entry_box.value)])
        set_task_added(True)

def clear_tasks():
    set_tasks(lambda v: [task for task in v if not task.done])

add_task_button = mo.ui.button(
    label="add task",
    on_change=lambda _: add_task(),
)

clear_tasks_button = mo.ui.button(
    label="clear completed tasks",
    on_change=lambda _: clear_tasks()
)
```
```
task_list = mo.ui.array(
    [mo.ui.checkbox(value=task.done, label=task.name) for task in get_tasks()],
    label="tasks",
    on_change=lambda v: set_tasks(
        lambda tasks: [Task(task.name, done=v[i]) for i, task in enumerate(tasks)]
    ),
)
```

It was hard to understand because for some examples on on_change, it's `lambda v:` while some other examples use `lambda _:`.
**When do we use `lambda v:`, when do we use `lambda _:`, and when do we use both?**

**What does v in  `on_change=lambda v: set_tasks` represent? Is it a list of booleans from the array of checkboxes?**

I tried to debug by adding prints into the `on_change`
```
def debug_type(v):
    print(type(v[0]))
    lambda v: set_tasks(
        lambda tasks: [
            Task(task.name, done=v[i]) for i, task in enumerate(tasks)
        ]
    )
    
task_list = mo.ui.array(
    [
        mo.ui.checkbox(value=task.done, label=task.name)
        for task in get_tasks()
    ],
    on_change=debug_type
)
```
**Why does this make the array only add tasks but can't clear anymore? Even removing the print (but still wrapping in `debug_type` does not clear when clicking clear.**

To understand what `i` is doing, i tried enumerating from 1 instead of default 0, and expected it to break which it did because i guess `v` is limited by the number of checkboxes in the array.
`Task(task.name, done=v[i]) for i, task in enumerate(tasks,1)`.

**However I don't understand why it only broke when I clicked the checkbox, and nothing happens when I click clear?  (i thought clear should break it too since clear is somehow part of this system)** 

**Please check if my understanding of the dataflow is correct.** I'm assuming `add task` goes through the same flow so i'll describe the more confusing `clear completed tasks`. 
I'm looking for something like this in the docs:

1. Clicking `clear completed tasks` triggers `set_tasks(lambda v: [task for task in v if not task.done])`  of `def clear_tasks()`
2. Calling `set_tasks` causes all cells with `get_tasks` of `mo.state([])` to run
3. `mo.ui.checkbox(value=task.done, label=task.name) for task in get_tasks()` contains get_tasks, so it ran to update the array UI
4. Because UI got updated, it caused `on_change=lambda v: set_tasks` to run (So this is the 2nd time `set_tasks` is running in this chain)
5. `get_tasks` in step 3 is not run again because it appears in the same cell as `on_change=lambda v: set_tasks`, so the loop ends

**Lastly, why is `task_list_mutated, set_task_list_mutated = mo.state(False)` and `set_task_list_mutated(True)` appearing in the handlers of the 2 buttons if `task_list_mutated` was never used? What's the point of setting if nothing is reading this state?**

I deleted this boolean state and all setting functions, and can't see any difference in behavior for both buttons in this whole task management UI. 

### Your Suggestion for Changes

- Suggest adding an explanation of the dataflow, best if in diagram form so users understand how `on_change` and `setters` interact, and how the loop ends when `set_tasks` is called twice from a single `clear completed tasks` trigger.
- Suggest deleting the boolean state if it does not contribute to the example.

**Comment by akshayka** [comment]
Regarding your simple examples such as

> To really stretch this example, i tried nesting composite elements, and saw randomly 1-5 rows of prints appended each time I toggled a checkbox.

I could not reproduce multiple rows being printed. For me only one row is printed on each interaction. This is my code: https://marimo.app/l/3fh9l9.

What operating system, marimo version, etc are you using? You can use `marimo env` at the command line to get this information.

> What is considered on_change in the mo.ui.array?

`on_change` is an "optional callback to run when this element’s value changes". So any time the value of the array changes, `on_change` should be run.

**Comment by akshayka** [comment]
In general we strongly discourage users from using `mo.state()`, which is difficult to reason about, and instead encourage them to rely on marimo's static analysis based reactivity. We plan to remove the state section from the UI tutorial.

**Comment by gitgithan** [comment]
I'm using wsl2 on windows.
`marimo env` shows
```
{
  "marimo": "0.9.14",
  "OS": "Linux",
  "OS Version": "5.15.153.1-microsoft-standard-WSL2",
  "Processor": "x86_64",
  "Python Version": "3.10.15",
  "Binaries": {
    "Browser": "--",
    "Node": "v20.15.1"
  },
  "Dependencies": {
    "click": "8.1.7",
    "docutils": "0.16",
    "itsdangerous": "2.2.0",
    "jedi": "0.19.1",
    "markdown": "3.6",
    "narwhals": "1.11.0",
    "packaging": "24.1",
    "psutil": "5.9.8",
    "pygments": "2.18.0",
    "pymdown-extensions": "10.11.2",
    "pyyaml": "6.0.1",
    "ruff": "0.7.1",
    "starlette": "0.27.0",
    "tomlkit": "0.12.5",
    "typing-extensions": "4.12.0",
    "uvicorn": "0.32.0",
    "websockets": "12.0"
  },
  "Optional Dependencies": {
    "altair": "5.4.1",
    "duckdb": "1.1.2",
    "pandas": "2.0.0",
    "polars": "1.7.0",
    "pyarrow": "18.0.0"
  }
}
```

I tried your example with only a single layer of `mo.ui.array` instead of my original example with nested `mo.ui.array((mo.ui.array...` and toggled the 1st checkbox. 
It still gave repeated outputs of 2,2,2,1,1 row each time.

```
[True, False]
[True, False]
[False, False]
[False, False]
[True, False]
[True, False]
[False, False]
[True, False]
```

> any time the value of the array changes, on_change should be run

This implies `on_change` should be trigged by not just toggling checkbox, but also by changes in length of task list from clicking the 2 buttons.
Thinking again, if clicking buttons causes the entire cell to be re-ran and `mo.ui.array` to be recreated, then even if `on_change` should respond to changes in length of task list, it may have already been overwritten by the cell rerun and the old reference to `task_list` is lost. 
So I can't see this on screen if `on_change` is also triggered by changes in number of tasks in the list.

**A failed experiment to prove that on_change reacts to list length changes :**
---
To prevent the whole `mo.ui.array` from being recreated and only change its value while keeping the same `mo.ui.array` instance, I tried to only update the value of the array UI element doing `complex.value = [mo.ui.checkbox() for _ in get_tasks_simple()]`, but this caused `RuntimeError: Setting the value of a UIElement is not allowed. If you need to imperatively set the value of a UIElement, consider using mo.state().`

```
get_tasks_simple, set_tasks_simple = mo.state([])
```
```
complex = mo.ui.array(
    (mo.ui.checkbox() for _ in range(2)), on_change=lambda v: print(v)
)
```
```
complex.value = [mo.ui.checkbox() for _ in get_tasks_simple()]
```
```
def add_task_simple():
    set_tasks_simple(lambda v: v + ['new hardcoded task'])
        
add_task_button_simple = mo.ui.button(
    label="add task",
    on_change=lambda _: add_task_simple(),
)
```
```
mo.hstack([complex, add_task_button_simple])
```

**Comment by liquidcarbon** [comment]
I'm also a bit confused about when (not) to use state.

Extending on the [recipe](https://github.com/marimo-team/marimo/blob/main/docs/recipes.md) about buttons, I'm trying to have a form-like UI without using a form, where input, button, and output are grouped together.

I want the button to trigger the action, but I want to be able to modify the input without erasing previous output.

Is this achievable without State?

![Image](https://github.com/user-attachments/assets/f638570b-ffec-412a-8586-81647ea6c91a)



```
# cell 1
text, button, TEXT

# cell 2
TEXT="initial"
if button.value:
    TEXT = text.value.upper()

# cell 3
button = mo.ui.run_button(label="CAPITALIZE")
text = mo.ui.text()
```

**Comment by akshayka** [comment]
@liquidcarbon Yes, this is possible, by judiciously mutating a regular Python object -- marimo doesn't track mutations. I find this to be much easier to reason about then using reactive state, since reactive state causes "action at a distance", forcing cells to re-run in a way that is not captured by the graph. This is the pattern that I follow in most of my more complex notebooks.

![Image](https://github.com/user-attachments/assets/500d3957-9bd1-43a0-860a-dfa6e5a0f24b)


```python
import marimo

__generated_with = "0.9.20"
app = marimo.App(width="medium")


@app.cell
def __():
    import marimo as mo
    return (mo,)


@app.cell
def __(TEXT, button, text):
    # cell 1
    text, button, TEXT[0]
    return


@app.cell
def __():
    # cell 2
    TEXT = ["initial"]
    return (TEXT,)


@app.cell
def __(TEXT, button, text):
    if button.value:
        TEXT[0] = text.value.upper()
    return


@app.cell
def __(mo):
    # cell 3
    button = mo.ui.run_button(label="CAPITALIZE")
    text = mo.ui.text()
    return button, text


if __name__ == "__main__":
    app.run()
```

**Comment by akshayka** [comment]
@liquidcarbon Yes, this is possible without state, by making use of the fact that Python does not mutate variables. I personally find state confusing because it causes cells to re-run in a way that is not captured by the graph, and prefer mutating regular Python variables instead -- that way I can still reason about when cells will re-run by looking at the dependency graph.

![Image](https://github.com/user-attachments/assets/8dbf4563-1def-4a2d-8134-fbd8626ba639)

```python
import marimo

__generated_with = "0.9.20"
app = marimo.App(width="medium")


@app.cell
def __():
    import marimo as mo
    return (mo,)


@app.cell
def __(TEXT, button, text):
    if button.value:
        TEXT[0] = text.value.upper()
        
    text, button, TEXT[0]
    return


@app.cell
def __():
    # create a container that holds the capitalized text
    # this cell won't ever re-run
    TEXT = ["initial"]
    return (TEXT,)


@app.cell
def __(mo):
    # cell 3
    button = mo.ui.run_button(label="CAPITALIZE")
    text = mo.ui.text()
    return button, text


if __name__ == "__main__":
    app.run()
```

**Comment by liquidcarbon** [comment]
This is great, why don't I add this to the recipes?

**Comment by akshayka** [comment]
Please do! :)
